Qt: extend OpenSSL collection for QtNetwork & implement strict Qt/QML plugin dependency validation

rokm opened the pull request:
Part 1: extend OpenSSL collection for QtNetwork
-----------------------------------------------

The first part of this PR extends the collection of OpenSSL shared libraries for `QtNetwork`, to handle different OpenSSL versions - 1.0.2 for very old PyQt5 wheels, 1.1.x for PySide2 and PyQt5, and >= 3.x for PySide6 and PyQt6 (and sometimes for Qt5, if using custom build). We also extend the collection, which was previously done only on Windows, to Linux (where PyPI wheels for all four bindings expect the shared lib to be provided by the environment) and macOS (where `openssl` TLS backend in Qt6 is enabled if OpenSSL dylib is available in the environment, e.g. from Homebrew).

To make collected OpenSSL dylib visible to `QtNetwork` in macOS POSIX builds, we need to add `sys._MEIPASS` to `DYLD_LIBRARY_PATH`, so have the run-time hook do that. This does not seem to be necessary for .app bundles, because `Contents/Frameworks` is automatically searched by `QtNetwork`.

Extend the `QtNetwork` OpenSSL test for Qt6, where we need to check if `openssl` TLS backend is available, instead of just checking for SSL/TLS support.

Part 2: implement strict Qt/QML plugin dependency validation
------------------------------------------------------------

The second part implements strict dependency validation for Qt plugins and for QML plugins.

Inside our Qt hook helpers, we now perform binary dependency analysis of Qt plugins, and discard the plugins that have missing Qt dependencies - for example, PyPI wheels ship Qt plugins with Qt dependencies that are missing from the wheels (they are missing completely, or require installation of an optional wheel). Such plugins are non-functional, so there is no point in collecting them. Plus, the missing dependencies would raise warnings later on during general binary dependency analysis. Same applies to the QML plugins.

Qt plugins with Qt dependencies not shipped by PyPI wheels have an even darker side - those dependencies might be available somewhere else in the system; for example, the underlying Linux distribution, a Homebrew environment on macOS, or an arbitrary Qt build that happens to be in `PATH` on Windows (see #8087). Even if they are of compatible versions, the Qt libs shipped with PyPI wheels and the external ones might not be fully compatible (e.g., different symbol visibility on Linux), so mixing the two is never a good idea.

That's why we also ensure that resolved Qt dependencies of plugins reside in the Qt shared library directory, as declared by `QLibraryInfo.PrefixPath` / `QLibraryInfo.BinariesPath` /  `QLibraryInfo.LibrariesPath` (whichever is applicable to the OS and wheel/non-wheel), and discard plugins with Qt dependencies that are resolved outside of that directory (note: "Qt dependencies" means Qt shared libraries - `Qt*.dll` or `libQt*.{so, dylib}`, as opposed to lower-level dependencies that might need to be collected from the host system).

For QML plugins, we now need to scan the QML directory for `qmldir` files to identify individual components, and parse them for lists of corresponding plugins (usually one plugin per component, though).

Examples of the new system in action, on Windows (without "fall-back" Qt installation in `PATH`):

<details><summary>PySide2</summary><p>

On Windows, `PySide2` does not seem to include any problematic plugins.

</p></details>

<details><summary>PySide6</summary><p>

With `PySide6` (which installed both `PySide6-Essential` and `PySide6-AddOns`), we now get the following warnings:
```
7156 INFO: Loading module hook 'hook-PySide6.QtQml.py' from '[...]\\pyinstaller\\PyInstaller\\hooks'...
13303 WARNING: QtLibraryInfo(PySide6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PySide6\\qml\\QtQuick3D\\Helpers\\impl\\qtquick3dhelpersimplplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DHelpersImpl.dll'.
13303 WARNING: QtLibraryInfo(PySide6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PySide6\\qml\\QtQuick3D\\Helpers\\impl' due to invalid plugin binaries!
13767 WARNING: QtLibraryInfo(PySide6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PySide6\\qml\\QtQuick3D\\SpatialAudio\\quick3dspatialaudioplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DSpatialAudio.dll'.
13767 WARNING: QtLibraryInfo(PySide6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PySide6\\qml\\QtQuick3D\\SpatialAudio' due to invalid plugin binaries!
```

</p></details>


<details><summary>PyQt5</summary><p>

With just `PyQt5` and `PyQtWebEngine` installed, we now get the following warnings:

```
6169 INFO: Loading module hook 'hook-PyQt5.QtQml.py' from '[...]\\PyInstaller\\hooks'...
6613 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtMultimedia\\declarative_multimedia.dll'! Reason: Missing Qt dependency 'Qt5MultimediaQuick.dll'.
6613 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtMultimedia' due to invalid plugin binaries!
7021 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtQuick\\Scene2D\\qtquickscene2dplugin.dll'! Reason: Missing Qt dependency 'Qt53DQuickScene2D.dll'.
7021 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtQuick\\Scene2D' due to invalid plugin binaries!
7055 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtQuick\\Scene3D\\qtquickscene3dplugin.dll'! Reason: Missing Qt dependency 'Qt53DLogic.dll'.
7055 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtQuick\\Scene3D' due to invalid plugin binaries!
```

If we install extra packages (`PyQt3D`, in particular), the warnings are reduced to:

```
6702 INFO: Loading module hook 'hook-PyQt5.QtQml.py' from '[...]\\PyInstaller\\hooks'...
8129 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtMultimedia\\declarative_multimedia.dll'! Reason: Missing Qt dependency 'Qt5MultimediaQuick.dll'.
8129 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt5\\Qt5\\qml\\QtMultimedia' due to invalid plugin binaries!
```

</p></details>

<details><summary>PyQt6</summary><p>

With just `PyQt6` and `PyQt6-WebEngine` installed, we now get the following warnings:

Similarly, for PyQt6 (`PyQt6` and `PyQt6-WebEngine` only):

```
5681 INFO: Loading module hook 'hook-PyQt6.QtQml.py' from '[...]\\PyInstaller\\hooks'...
6907 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQml\\XmlListModel\\qmlxmllistmodelplugin.dll'! Reason: Missing Qt dependency 'Qt6QmlXmlListModel.dll'.
6907 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQml\\XmlListModel' due to invalid plugin binaries!
8052 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Effects\\effectsplugin.dll'! Reason: Missing Qt dependency 'Qt6QuickEffects.dll'.
8052 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Effects' due to invalid plugin binaries!
8159 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\LocalStorage\\qmllocalstorageplugin.dll'! Reason: Missing Qt dependency 'Qt6QmlLocalStorage.dll'.
8160 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\LocalStorage' due to invalid plugin binaries!
8423 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Scene2D\\qtquickscene2dplugin.dll'! Reason: Missing Qt dependency 'Qt63DRender.dll'.
8423 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Scene2D' due to invalid plugin binaries!
8701 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Scene3D\\qtquickscene3dplugin.dll'! Reason: Missing Qt dependency 'Qt63DLogic.dll'.
8701 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Scene3D' due to invalid plugin binaries!
9197 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Helpers\\impl\\qtquick3dhelpersimplplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DHelpersImpl.dll'.
9197 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Helpers\\impl' due to invalid plugin binaries!
9319 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\ParticleEffects\\qtquick3dparticleeffectsplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DParticleEffects.dll'.
9320 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\ParticleEffects' due to invalid plugin binaries!
9428 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics\\Helpers\\qtquick3dphysicshelpersplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DPhysicsHelpers.dll'.
9428 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics\\Helpers' due to invalid plugin binaries!
9475 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics\\qquick3dphysicsplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DPhysics.dll'.
9475 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics' due to invalid plugin binaries!
```

Installing `PyQt6-3D` fixes some of the missing dependencies, but also brings additional plugins with missing dependencies:

```
6668 INFO: Loading module hook 'hook-PyQt6.QtQml.py' from '[...]\\PyInstaller\\hooks'...
6930 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\Qt3D\\Core\\quick3dcoreplugin.dll'! Reason: Missing Qt dependency 'Qt63DQuick.dll'.
6930 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\Qt3D\\Core' due to invalid plugin binaries!
6995 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\Qt3D\\Render\\quick3drenderplugin.dll'! Reason: Missing Qt dependency 'Qt63DQuick.dll'.
6995 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\Qt3D\\Render' due to invalid plugin binaries!
7093 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQml\\XmlListModel\\qmlxmllistmodelplugin.dll'! Reason: Missing Qt dependency 'Qt6QmlXmlListModel.dll'.
7093 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQml\\XmlListModel' due to invalid plugin binaries!
7373 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Effects\\effectsplugin.dll'! Reason: Missing Qt dependency 'Qt6QuickEffects.dll'.
7373 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\Effects' due to invalid plugin binaries!
7404 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\LocalStorage\\qmllocalstorageplugin.dll'! Reason: Missing Qt dependency 'Qt6QmlLocalStorage.dll'.
7404 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick\\LocalStorage' due to invalid plugin binaries!
7616 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Helpers\\impl\\qtquick3dhelpersimplplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DHelpersImpl.dll'.
7616 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Helpers\\impl' due to invalid plugin binaries!
7648 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\ParticleEffects\\qtquick3dparticleeffectsplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DParticleEffects.dll'.
7648 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\ParticleEffects' due to invalid plugin binaries!
7665 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics\\Helpers\\qtquick3dphysicshelpersplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DPhysicsHelpers.dll'.
7665 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics\\Helpers' due to invalid plugin binaries!
7681 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics\\qquick3dphysicsplugin.dll'! Reason: Missing Qt dependency 'Qt6Quick3DPhysics.dll'.
7681 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]\\venv\\Lib\\site-packages\\PyQt6\\Qt6\\qml\\QtQuick3D\\Physics' due to invalid plugin binaries!
```

</p></details>


On my Fedora linux (which has Qt5 and Qt6 installed as part of the system):

<details><summary>PySide6</summary><p>

```
3454 INFO: Loading module hook 'hook-PySide6.QtQml.py' from '[...]/PyInstaller/hooks'...
4535 WARNING: QtLibraryInfo(PySide6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PySide6/Qt/qml/QtQuick3D/Helpers/impl/libqtquick3dhelpersimplplugin.so'! Reason: Qt dependency 'libQt6QmlModels.so.6' ('/usr/lib64/libQt6QmlModels.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PySide6/Qt/lib').
4535 WARNING: QtLibraryInfo(PySide6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PySide6/Qt/qml/QtQuick3D/Helpers/impl' due to invalid plugin binaries!
4586 WARNING: QtLibraryInfo(PySide6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PySide6/Qt/qml/QtQuick3D/SpatialAudio/libquick3dspatialaudioplugin.so'! Reason: Missing Qt dependency 'libQt6Quick3DSpatialAudio.so.6'.
4587 WARNING: QtLibraryInfo(PySide6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PySide6/Qt/qml/QtQuick3D/SpatialAudio' due to invalid plugin binaries!
```

</p></details>


<details><summary>PyQt5</summary><p>

```
3291 INFO: Loading module hook 'hook-PyQt5.QtGui.py' from '[...]/PyInstaller/hooks'...
3359 WARNING: QtLibraryInfo(PyQt5): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt5/Qt5/plugins/egldeviceintegrations/libqeglfs-x11-integration.so' ('egldeviceintegrations')! Reason: Qt dependency 'libQt5EglFSDeviceIntegration.so.5' ('/usr/lib64/libQt5EglFSDeviceIntegration.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
3371 WARNING: QtLibraryInfo(PyQt5): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt5/Qt5/plugins/egldeviceintegrations/libqeglfs-kms-egldevice-integration.so' ('egldeviceintegrations')! Reason: Qt dependency 'libQt5EglFsKmsSupport.so.5' ('/usr/lib64/libQt5EglFsKmsSupport.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
3381 WARNING: QtLibraryInfo(PyQt5): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt5/Qt5/plugins/egldeviceintegrations/libqeglfs-emu-integration.so' ('egldeviceintegrations')! Reason: Qt dependency 'libQt5EglFSDeviceIntegration.so.5' ('/usr/lib64/libQt5EglFSDeviceIntegration.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
3569 WARNING: QtLibraryInfo(PyQt5): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt5/Qt5/plugins/platforms/libqeglfs.so' ('platforms')! Reason: Qt dependency 'libQt5EglFSDeviceIntegration.so.5' ('/usr/lib64/libQt5EglFSDeviceIntegration.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
...
4351 INFO: Loading module hook 'hook-PyQt5.QtQml.py' from '[...]/PyInstaller/hooks'...
4570 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/qml/QtMultimedia/libdeclarative_multimedia.so'! Reason: Qt dependency 'libQt5MultimediaQuick.so.5' ('/usr/lib64/libQt5MultimediaQuick.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
4570 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/qml/QtMultimedia' due to invalid plugin binaries!
4824 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/qml/QtQuick/Scene2D/libqtquickscene2dplugin.so'! Reason: Qt dependency 'libQt53DLogic.so.5' ('/usr/lib64/libQt53DLogic.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
4824 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/qml/QtQuick/Scene2D' due to invalid plugin binaries!
4833 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/qml/QtQuick/Scene3D/libqtquickscene3dplugin.so'! Reason: Qt dependency 'libQt53DLogic.so.5' ('/usr/lib64/libQt53DLogic.so.5.15.11') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/lib').
4833 WARNING: QtLibraryInfo(PyQt5): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt5/Qt5/qml/QtQuick/Scene3D' due to invalid plugin binaries!
```

</p></details>


<details><summary>PyQt6</summary><p>

```
3198 INFO: Loading module hook 'hook-PyQt6.QtGui.py' from '[...]/PyInstaller/hooks'...
3422 WARNING: QtLibraryInfo(PyQt6): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt6/Qt6/plugins/platforms/libqeglfs.so' ('platforms')! Reason: Qt dependency 'libQt6EglFSDeviceIntegration.so.6' ('/usr/lib64/libQt6EglFSDeviceIntegration.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
3978 WARNING: QtLibraryInfo(PyQt6): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt6/Qt6/plugins/egldeviceintegrations/libqeglfs-x11-integration.so' ('egldeviceintegrations')! Reason: Qt dependency 'libQt6EglFSDeviceIntegration.so.6' ('/usr/lib64/libQt6EglFSDeviceIntegration.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4071 WARNING: QtLibraryInfo(PyQt6): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt6/Qt6/plugins/egldeviceintegrations/libqeglfs-kms-egldevice-integration.so' ('egldeviceintegrations')! Reason: Qt dependency 'libQt6EglFsKmsSupport.so.6' ('/usr/lib64/libQt6EglFsKmsSupport.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4091 WARNING: QtLibraryInfo(PyQt6): excluding plugin '[...]/venv/lib64/python3.10/site-packages/PyQt6/Qt6/plugins/egldeviceintegrations/libqeglfs-emu-integration.so' ('egldeviceintegrations')! Reason: Qt dependency 'libQt6EglFSDeviceIntegration.so.6' ('/usr/lib64/libQt6EglFSDeviceIntegration.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
...
4349 INFO: Loading module hook 'hook-PyQt6.QtQml.py' from '[...]/PyInstaller/hooks'...
4514 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQml/XmlListModel/libqmlxmllistmodelplugin.so'! Reason: Qt dependency 'libQt6QmlXmlListModel.so.6' ('/usr/lib64/libQt6QmlXmlListModel.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4514 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQml/XmlListModel' due to invalid plugin binaries!
4772 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/Effects/libeffectsplugin.so'! Reason: Qt dependency 'libQt6QuickEffects.so.6' ('/usr/lib64/libQt6QuickEffects.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4772 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/Effects' due to invalid plugin binaries!
4803 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/LocalStorage/libqmllocalstorageplugin.so'! Reason: Qt dependency 'libQt6QmlLocalStorage.so.6' ('/usr/lib64/libQt6QmlLocalStorage.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4803 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/LocalStorage' due to invalid plugin binaries!
4880 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D/libqtquickscene2dplugin.so'! Reason: Qt dependency 'libQt63DQuickScene2D.so.6' ('/usr/lib64/libQt63DQuickScene2D.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4880 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D' due to invalid plugin binaries!
4906 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.so'! Reason: Qt dependency 'libQt63DAnimation.so.6' ('/usr/lib64/libQt63DAnimation.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
4906 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D' due to invalid plugin binaries!
5025 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/Helpers/impl/libqtquick3dhelpersimplplugin.so'! Reason: Qt dependency 'libQt6OpenGL.so.6' ('/usr/lib64/libQt6OpenGL.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
5025 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/Helpers/impl' due to invalid plugin binaries!
5057 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/ParticleEffects/libqtquick3dparticleeffectsplugin.so'! Reason: Qt dependency 'libQt6Quick3DParticleEffects.so.6' ('/usr/lib64/libQt6Quick3DParticleEffects.so.6.6.0') has been resolved outside of the Qt shared library directory ('[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/lib').
5057 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/ParticleEffects' due to invalid plugin binaries!
5085 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/Physics/Helpers/libqtquick3dphysicshelpersplugin.so'! Reason: Missing Qt dependency 'libQt6Quick3DPhysicsHelpers.so.6'.
5085 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/Physics/Helpers' due to invalid plugin binaries!
5100 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin binary '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/Physics/libqquick3dphysicsplugin.so'! Reason: Missing Qt dependency 'libQt6Quick3DPhysics.so.6'.
5100 WARNING: QtLibraryInfo(PyQt6): excluding QML plugin directory '[...]/venv/lib/python3.10/site-packages/PyQt6/Qt6/qml/QtQuick3D/Physics' due to invalid plugin binaries!
```
</p></details>

rokm committed:
hookutils: qt: pre-compute the resolved Qt shared libs path

Extend QtLibraryInfo with the following properties:

- package_location: fully-resolved pathlib.Path() pointing to the
  bindings' python package location. All helpers that were using
  old string-based package_location were turning it into
  fully-resolved pathlib.Path themselves.

- qt_inside_package: a flag indicating that Qt is bundled within
  the bindings' python package (i.e., PyPI wheels).

- qt_lib_dir: a fully-resolved pathlib.Path() pointing to the
  directory that contains Qt shared libraries. This abstracts the
  difference between general Windows and non-Windows case, where
  `BinariesPath` and `LibrariesPath` location are used, respectively.
  It also handles the PySide2/6 PyPI wheels for Windows, where
  the `PrefixPath` is used instead.

Adjust the existing helpers to use these pre-computed properties
where possible.

rokm committed:
hookutils: qt: extend the search for OpenSSL DLLs on Windows

Extend the search for OpenSSL DLLs for QtNetwork to cover both
32-bit and 64-bit variant (depending on the environment), and
to cover OpenSSL 3.0.x that is used with contemporary Qt6 (and
in some cases, even Qt5).

We now use `bindepend.resolve_library_path` to search for the
DLLs, so we search the default Qt shared libraries directory and
fall to additional search paths (e.g., PATH).

rokm committed:
hookutils: qt: prefer package-provided OpenSSL DLLs over external ones

When collecting OpenSSL DLLs for QtNetwork, apply an additional
filter; if we found at least one OpenSSL within the bindings'
python package, discard all external OpenSSL DLLs, as they are
very likely redundant.

rokm committed:
tests: extend test_Qt_QtNetwork_SSL_support for Qt >= 6.1

Under Qt >= 6.1, we need to test for explicit availability of the
'openssl' TLS backend instead of just for SSL support being
available.

rokm committed:
hookutils: qt: move shared library normalization into a helper

This way, we will be able to re-use it when trying to determine
whether a dependency of a plugin is a Qt shared library or not.

rokm committed:
hookutils: qt: implement dependency validation for Qt plugins

Scan the dependency of Qt plugins and automatically exclude plugins
where:
- a Qt shared library dependency is missing
- a Qt shared library dependency is resolved outside of the
  Qt shared library directory

The first condition prevents us from collecting plugins that are
defunct in the build environment.

The second condition prevents plugins from missing Qt dependencies
from pulling those dependencies from unrelated Qt environments
that (in)conveniently happen to be in the search path.

For example, PyQt5 PyPI wheels ship several plugins with missing
dependencies (some of which are in optional add on packages),
and those might end up pulling in system-installed Qt libraries
(linux), Homebrew installation (macOS), or an arbitrary Qt
build that happens to be in PATH (Windows).

rokm committed:
tests: test_Qt_QtNetwork_SSL_support: increase verbosity of the test

rokm committed:
hookutils: qt: query OpenSSL version via QSslSocket method

Query the run-time version of OpenSSL used by QtNetwork, using the
`QSslSocket.sslLibraryVersionNumber` method. This allows us to
precisely determine the names of OpenSSL DLLs that we need to
search for on Windows.

rokm committed:
hookutils: qt: search and collect OpenSSL dylibs on macOS

Extend `collect_qtnetwork_files` to collect OpenSSL dynamic
libraries on macOS. After extending the QtNetwork test to check
whether `openssl` TLS backend is available, we can see failures
with python 3.8, 3.9, and 3.10 when running with latest PyQt6
and PySide6, which both use Qt 6.6.0. Those python versions likely
provide OpenSSL 1.1.x, while Qt was built with 3.0.x.

On the CI runner, OpenSSL 3.0.x is likely available from Homebrew,
so the tests sees the `openssl` TLS  backend available in the
unfrozen python + Qt, and thus is not skipped. When running the
frozen application, the POSIX build passes, but .app bundle does
not; because the former can access the Homebrew copy of OpenSSL
at run-time, while the latter cannot.

Which means that we should be attempting to collect the OpenSSL
dylibs for QtNetwork on macOS as well.

rokm committed:
rthooks: PyQt6/PySide6: add sys._MEIPASS to DYLD_LIBRARY_PATH

Have PyQt6 and PySide6 run-time hooks add sys._MEIPASS to
DYLD_LIBRARY_PATH when not running as a macOS app bundle. This
is required for QtNetwork in Qt6 to be able to pick up OpenSSL
dynamic libraries for its `openssl` TLS backend; this both ensures
that the backend is functional if OpenSSL libs were collected, and
also prevents fallback to external locations, such as Homebrew.

Applicable only to Qt6 and its bindings; macOS Qt5 builds seem to
use the OS-provided SecureTransport API instead of OpenSSL.

rokm committed:
hookutils: qt: implement OpenSSL collection on Linux

On Linux, we should also be collecting OpenSSL libraries if
required by QtNetwork's SSL support (built-in for Qt5, `openssl`
TLS backend in Qt6).

rokm committed:
hookutils: qt: raise OpenSSL version limit for 3.x series to 4.0

Looks like OpenSSL 3.1 is ABI-compatible with 3.0; so for now,
assume that the next ABI break will be at 4.0.

rokm committed:
hookutils: qt: implement dependency validation for QML plugins

Implement advanced processing of Qt QML plugin directories; we now
scan base QML plugin directory for `qmldir` files, and process each
directory that contains such a file as a separate QML plugin directory.

In particular, we scan the `qmldir` file for declared plugin binaries,
and then perform strict dependency validation; i.e., we discard
the binary if:
 - a Qt shared library dependency is missing
 - a Qt shared library dependency is resolved outside of the
   Qt shared library directory

The first condition prevents us from collecting plugins that are
defunct in the build environment. The second condition prevents plugins
from missing Qt dependencies from pulling those dependencies from
unrelated Qt environments that (in)conveniently happen to be in the
search path.

If any plugin binary from a given QML plugin directory has been
discarded, we completely ignore that plugin directory (i.e., we
completely ignore component with defunct plugin binary).

Otherwise, we collect the plugin binary as a binary file, and all
other files from the directory as data files. If the directory
contains sub-directories, we collect those that do not contain
`qmldir` files (as those who do are processed on their own).

rokm committed:
bindepend: fix error when calling get_imports() with pathlib.Path

Fix an error when calling `PyInstaller.depend.bindepend.get_imports`
with a `filename` that is a `pathlib.Path` (or rather, not a `str`)
on Windows. This particular error is caused by trying to call
`lower()` under assumption that `filename` is an instance of `str`;
so do the explicit cast to `str` for that part.

rokm committed:
rthooks: Qt: consolidate code that prepends sys._MEIPASS to environment variable

Consolidate the code that prepends `sys._MEIPASS` to environment
variable (`PATH` on Windows, `DYLD_LIBRARY_PATH` on macOS for
POSIX builds) into a new helper in the ˙_pyi_rth_utils` module,
called `prepend_path_to_environment_variable`.

Update the comments in Windows codepath to clarify why `QtNetwork`
requires `sys._MEIPASS` to be added to `PATH`.

bwoodsend commented:
> To make collected OpenSSL dylib visible to QtNetwork in macOS POSIX builds, we need to add sys._MEIPASS to DYLD_LIBRARY_PATH, so have the run-time hook do that

Is this really not going to have side effects? I'm imagining a different, unrelated dylib being loaded in response to `QtNetwork` being imported.

Edit: Reading the corresponding code change reminded me that we use `rpath` on macOS so presumably `DYLD_LIBRARY_PATH` should be limited in the damage it can cause.

bwoodsend commented:
Are you sure we want all those new warnings to be warnings? It sounds like people are going to be seeing a lot of them merely because they have say PySide6 installed.

bwoodsend commented on PyInstaller/utils/hooks/qt/__init__.py:
```diff
@@ -809,6 +812,50 @@ def _collect_qtnetwork_openssl_macos(self, openssl_version):
+        Linux-specific collection of OpenSSL dylibs required by QtNetwork module.
+        """
+
+        # Out of the supported OSes, Linux is by most straight-forward, because OpenSSL shared libraries are expected
```
I'm guessing there should be a `far the` after that `by`?

rokm replied:
Indeed.

bwoodsend commented on PyInstaller/utils/hooks/qt/__init__.py:
```diff
@@ -525,7 +525,46 @@ def collect_plugins(self, plugin_type):
+        """
+
+        imported_libraries = bindepend.get_imports(plugin_file, search_paths=[self.qt_lib_dir])
+        while imported_libraries:
+            imported_lib_name, imported_lib_path = imported_libraries.pop()  # (name, fullpath) tuple
```
Why can't this just be a generic 
```python
for (imported_lib_name, imported_lib_path) in imported_libraries:

rokm replied:
It can, and it should.

rokm commented:
> > To make collected OpenSSL dylib visible to QtNetwork in macOS POSIX builds, we need to add sys._MEIPASS to DYLD_LIBRARY_PATH, so have the run-time hook do that
> 
> Is this really not going to have side effects? I'm imagining a different, unrelated dylib being loaded in response to `QtNetwork` being imported.
> 
> Edit: Reading the corresponding code change reminded me that we use `rpath` on macOS so presumably `DYLD_LIBRARY_PATH` should be limited in the damage it can cause.

It should not affect the main process, because similarly to `LD_LIBRARY_PATH`, `DYLD_LIBRARY_PATH` needs to be set before the process is ran in order to have any effect. That's actually the main reason why we use `rpath`, because otherwise we would need two processes even for `onedir` (like we used to have for Linux), or have the process restart itself after modifying the environment (like we do now for Linux); I imagine the former approach was problematic because it resulted in two application icons in the UI, and I suspect the latter would cause all sort of problems with app bundles.

It might affect the search path in subprocesses, but we have `rpaths` pointing to `sys._MEIPASS` anyway, so I don't expect much difference in that regard.

But it will have potential side effects when trying to run external programs from the frozen application - which is why we now have https://pyinstaller.org/en/latest/common-issues-and-pitfalls.html#macos

---

If we ever get around to implementing run-time hooks that are triggered on module import, we could have such on-demand `PyQt6` or `PySide6` hook modify `DYLD_LIBRARY_PATH`, import `QtNetwork` module (and run `QtNetwork.QSslSocket.supportsSsl()`), and then restore `DYLD_LIBRARY_PATH`. Same with `PATH` on Windows.

While we could also do that now, I would prefer not to, because currently we run all run-time hooks in `multiprocessing` worker processes as well, so importing `PyQt6` or `PySide6` would unnecessarily slow those down.

rokm commented:
> Are you sure we want all those new warnings to be warnings? It sounds like people are going to be seeing a lot of them merely because they have say PySide6 installed.

Hmm, yeah, originally I wanted them to be warnings. So that people would see them, and complain to `PyQt6` / `PySide6` to include the missing Qt libs. But obviously what is going to happen is that people are going to complain to *us* about the new warnings. I'll demote them to debug messages instead.

rokm committed:
hookutils: qt: demote warnings from plugin validation to debug messages

Demote warnings generated by Qt plugin and QML component/plugin
dependency validation to debug messages.
